### PR TITLE
app, controller: allow recurring backup for detached volumes

### PR DESCRIPTION
**Pull request by PhanLe1010**
In the current Longhorn implementation, users cannot do recurring backup when volumes are detached. 
This PR gives the users an option to do recurring backup even when volumes are detached.

The design overview is documented at the PR https://github.com/longhorn/longhorn/pull/1851

longhorn/longhorn#1509

### CODE REVIEW

File: client/generated_volume.go [hunk 1]
```diff
@@ -120,6 +120,8 @@ type VolumeOperations interface {
 	ActionSnapshotPurge(*Volume) (*Volume, error)
 
 	ActionSnapshotRevert(*Volume, *SnapshotInput) (*Snapshot, error)
+
+	ActionUpdateDataLocality(*Volume, *UpdateDataLocalityInput) (*Volume, error)
```
Comment (yasker) [inline client/generated_volume.go:124]: I don't think we need this in this PR? If we still want the action, we can update it in another PR.

File: api/model.go [hunk 2]
```diff
@@ -537,7 +537,8 @@ func volumeSchema(volume *client.Schema) {
 		},
 
 		"updateDataLocality": {
-			Input: "UpdateDataLocalityInput",
+			Input:  "UpdateDataLocalityInput",
```
Comment (yasker) [inline api/model.go:540]: Unrelated change.

File: app/volume.go [hunk 3]
```diff
@@ -31,6 +33,10 @@ const (
 	SnapshotPurgeStatusInterval = 5 * time.Second
 
 	WaitInterval = 5 * time.Second
+	timeout      = 300 // 5 minutes
```
Comment (yasker) [inline app/volume.go:36]: Give `timeout` a name, e.g. `VolumeAttachTimeout`.

File: app/volume.go [hunk 4]
```diff
@@ -41,6 +47,10 @@ func SnapshotCmd() cli.Command {
 				Name:  FlagManagerURL,
 				Usage: "Longhorn manager API URL",
 			},
+			cli.StringFlag{
+				Name:  FlagCurrentNode,
```
Comment (yasker) [inline app/volume.go:51]: Why do we need this?

File: app/volume.go [hunk 5]
```diff
@@ -274,12 +346,35 @@ func (job *Job) backupAndCleanup() (err error) {
...
 
+	volume, err := volumeAPI.ById(volumeName)
+	if err != nil {
+		return errors.Wrapf(err, "could not get volume %v", volumeName)
+	}
+
+	// Check the size of VolumeHead, if it is empty, skip the recurring backup.
+	// We don't want to overwrite the old backups with new identical backups.
+	// This happens when the volume is detached for a long time and has no new data.
+	volumeHead, err := volumeAPI.ActionSnapshotGet(volume, &longhornclient.SnapshotInput{
```
Comment (yasker) [inline app/volume.go:363]: Separate the volume head size check into another commit. It will affect all the recurring jobs, not only detached one.

File: app/volume.go [hunk 6]
```diff
@@ -386,3 +481,21 @@ func (job *Job) GetVolume(name string) (*longhorn.Volume, error) {
 func (job *Job) UpdateVolumeStatus(v *longhorn.Volume) (*longhorn.Volume, error) {
 	return job.lhClient.LonghornV1beta1().Volumes(job.namespace).UpdateStatus(v)
 }
+
+// waitForVolumeToBeInState timeout in second
+func (job *Job) waitForVolumeToBeInState(state string, timeout int) (*longhornclient.Volume, error) {
```
Comment (yasker) [inline app/volume.go:486]: just call it `waitForVolumeState`

File: app/volume.go (hunk 4 above)
Comment (yasker) [reply app/volume.go:51]: OK I see. But what if the specified node is down? Seems better if we randomly choose a ready node.

File: controller/volume_controller.go [hunk 7]
```diff
@@ -1940,7 +1943,8 @@ func (vc *VolumeController) createCronJob(v *longhorn.Volume, job *types.Recurri
 								},
 							},
 							ServiceAccountName: vc.ServiceAccount,
-							RestartPolicy:      v1.RestartPolicyOnFailure,
+							//RestartPolicy:      v1.RestartPolicyOnFailure,
+							RestartPolicy: v1.RestartPolicyNever,
```
Comment (yasker) [inline controller/volume_controller.go:1947]: ?

**Review by yasker — CHANGES_REQUESTED** [review]

File: api/model.go (hunk 2 above)
Comment (PhanLe1010) [reply api/model.go]: Thank you. Fixed

File: client/generated_volume.go (hunk 1 above)
Comment (PhanLe1010) [reply client/generated_volume.go]: Thank you. Fixed

File: app/volume.go (hunk 3 above)
Comment (PhanLe1010) [reply app/volume.go]: Thank you. Fixed

File: app/volume.go (hunk 4 above)
Comment (PhanLe1010) [reply app/volume.go]: Thank you. Fixed

File: app/volume.go (hunk 6 above)
Comment (PhanLe1010) [reply app/volume.go]: Thank you. Fixed

File: controller/volume_controller.go (hunk 7 above)
Comment (PhanLe1010) [reply controller/volume_controller.go]: My bad. I forgot to remove debugging code

File: app/volume.go [hunk 8]
```diff
@@ -386,3 +459,56 @@ func (job *Job) GetVolume(name string) (*longhorn.Volume, error) {
...
+				readyNodeList = append(readyNodeList, node.Name)
+			}
+		}
+	}
+
+	if len(readyNodeList) == 0 {
+		return "", fmt.Errorf("cannot find a ready node")
+	}
+
+	return readyNodeList[rand.Int()%len(readyNodeList)], nil
```
Comment (yasker) [inline app/volume.go:532]: Use `intn` instead.

**Review by yasker — CHANGES_REQUESTED** [review]

File: app/volume.go (hunk 8 above)
Comment (PhanLe1010) [reply app/volume.go]: Thank you. Fixed

**Review by yasker — APPROVED** [review]

**Comment by yasker** [comment]
LGTM. Waiting for the test result.
